lib/Basic/Platform.cpp

```diff
@@ -431,147 +431,139 @@ llvm::Triple swift::getUnversionedTriple(const llvm::Triple &triple) {
                       unversionedOSName);
 }
 
+namespace {
+
+enum class Platform {
+  macOS,
+  iOS,
+  watchOS,
+  visionOS
+};
+
+/// A specific platform release.
+struct PlatformRelease {
+  Platform platform;
+  llvm::VersionTuple version;
+};
+
+struct SwiftRelease {
+  /// The release version of the Swift runtime.
+  llvm::VersionTuple version;
+
+  /// Platform releases that provide this release of the Swift runtime.
+  std::initializer_list<PlatformRelease> providers;
+};
+
+} // end anonymous namespace
+
+// TODO: do some constexpr trickery to split these out by platform
+static constexpr SwiftRelease swiftReleases[] = {
+#define PLATFORM(NAME, VERSION)     \
+      {                             \
+        Platform::NAME,             \
+        llvm::VersionTuple VERSION  \
+      },
+#define FUTURE
+#define RUNTIME_VERSION(SWIFT_TUPLE, PROVIDERS) \
+  {                                             \
+    llvm::VersionTuple SWIFT_TUPLE,             \
+    {                                           \
+      PROVIDERS                                 \
+    }                                           \
+  },
+#include "swift/AST/RuntimeVersions.def"
+};
+
+/// Return the highest Swift release that matches the given platform and
+/// has a version no greater than the target version. Don't return a version
+/// older that the minimum. Returns null if the target version matches the
+/// notional future release version.
+static std::optional<llvm::VersionTuple>
+findSwiftRuntimeVersion(Platform targetPlatform,
+                        llvm::VersionTuple targetVersion,
+                        llvm::VersionTuple minimumRelease) {
+  #define MAX(a, b) ((a) > (b) ? (a) : (b))
+
+  // Just do a simple reverse scan for now.
+  for (auto &swiftRelease : llvm::reverse(llvm::ArrayRef(swiftReleases))) {
+    for (auto &provider : swiftRelease.providers) {
+      // If the provider version is <= the deployment target, then
+      // the deployment target includes support for the given Swift
+      // release. Since we're scanning in reverse order of Swift
+      // releases (because of the order of entries in RuntimeVersions.def),
+      // this must be the highest supported Swift release.
+      if (provider.platform == targetPlatform &&
+          provider.version <= targetVersion) {
+        return MAX(swiftRelease.version, minimumRelease);
```

Comment:
Can we use `std::max` instead?

Reply:
The existing code wasn't, and that seemed deliberate. I didn't look into it.

Reply:
Perhaps it was because this code was written before we had C++14 as the default language mode.

Reply:
Alright, I'll get it https://github.com/swiftlang/swift/pull/78819